linux/just_files/just_pre-commit_functions.bsh

```diff
@@ -0,0 +1,294 @@
+      unset hook_files[hook_index]
+    fi
+  done
+}
+
+
+function pre-commit()
+{
+  # this Does't work
+  # PRE_COMMIT_HOME=${precommit_dir}/cache
+
+  local subcommand=()
+  if (( ${#} )); then
+    case "${1}" in
+      autoupdate|init-templatedir|install|install-hooks|migrate-config|run|try-repo|uninstall)
+        # The commands are executed in some weird .cache dir, relative path to requirements will never work
+        sed 's|\${precommit_dir}|'"${precommit_dir}"'|g' "${precommit_dir}/config.yml.in" > "${precommit_dir}/config.yml"
+        subcommand=("${1}" -c "${precommit_dir}/config.yml")
+        shift 1
+        ;;
+    esac
+  fi
+
+  local JUST_IGNORE_EXIT_CODES=1
+  ${DRYRUN} "$(vsi::python::virtualenv_bin "${python_dir}" pre-commit)" \
+    ${subcommand[@]+"${subcommand[@]}"} ${@+"${@}"}
+
+  if (( ${#subcommand[@]} )) && [[ ${subcommand[0]} = install* ]]; then
+    local hook hook_files
+    pre-commit::_get_hook_files
+    for hook in ${hook_files[@]+"${hook_files[@]}"}; do
+      if ! grep -q pre-commit_filter_lines "${hook}"; then
+        echo "Patching $(basename "${hook}")..." >&2
+        # Designed to patch https://github.com/pre-commit/pre-commit/blob/main/pre_commit/resources/hook-tmpl
+        ${DRYRUN} sed -i 's|^ *exec.*pre[-_]commit "\${ARGS\[@]}"$|& < <("'"${VSI_COMMON_DIR}"'/linux/pre-commit_filter_lines")|' "${hook}"
+      fi
+    done
+  fi
+}
+
+
+function pre-commit::defaultify()
+{
+  local project_cwd="${JUST_PROJECT_PREFIX}_CWD"
+  project_cwd=${!project_cwd}
+  local precommit_dir="${JUST_PROJECT_PREFIX}_PRECOMMIT_DIR"
+  precommit_dir=${!precommit_dir:-${project_cwd}/.pre-commit}
+  local python_dir="${JUST_PROJECT_PREFIX}_PRECOMMIT_PYTHON_DIR"
+  python_dir=${!python_dir-${precommit_dir}/python}
+  local desired_python_version="${JUST_PROJECT_PREFIX}_PRECOMMIT_PYTHON_VERSION"
+  desired_python_version=${!desired_python_version-3.12.1}
+
+  local hook_types="${JUST_PROJECT_PREFIX}_PRECOMMIT_HOOKS[@]"
+  # This won't recognize an array set as empty as a set state, it will be treated as unset
+  hook_types=("${!hook_types-pre-push}")
+
+  arg="${1}"
+  shift 1
+  case ${arg} in
+    pre-commit) # Run pre-commit command (e.g. just pre-commit run)
+      pre-commit "${@}"
+      extra_args=${#}
+      ;;
+    pre-commit_run) # Run pre-commit check locally. Defaults to "-a" to run everything
+      if (( ${#} )); then
+        pre-commit run "${@}"
+        extra_args=${#}
+      else
+        pre-commit run -a
+      fi
+      ;;
+    pre-commit_init) # Initialize pre-commit for the first time in a new project
+      mkdir -p "${precommit_dir}"
+      cp "${VSI_COMMON_DIR}/templates/pre-commit/requirements.in" "${precommit_dir}/requirements.in"
+      cp "${VSI_COMMON_DIR}/templates/pre-commit/config.yml.in" "${precommit_dir}/config.yml.in"
+
+      local precommit_rel=$(relative_path "${precommit_dir}" "${project_cwd}")
+      local precommit_rel_regex=$(regex_escape "${precommit_rel}")
```

Comment:
add a comment that you are adding these files to the .gitignore if they are not there already

Reply:
```suggestion
      local precommit_rel_regex=$(regex_escape "${precommit_rel}")
      # Add to .gitignore if it is not in there already
```